### PR TITLE
Feature: Save Data Support

### PR DESCRIPTION
**Pull request by debughug**
**Screenshot of Change**
image

**Purpose**
Giving the board a chance to store the board on handleSquareClick events. It'll retain the data and place it on refresh of the browser using a cookie.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,14 +1,13 @@
 import React from 'react';
 import { GameBoard } from './Components/GameBoard';
-import { BrowserRouter, Route } from 'react-router-dom';
```
Comment (debughug) [inline src/App.js:3]: IDE defines this as never used - removing

File: src/Components/GameBoard.js [hunk 2]
```diff
@@ -1,17 +1,33 @@
 import React from 'react';
-import { Link } from 'react-router-dom';
```
Comment (debughug) [inline src/Components/GameBoard.js:2]: IDE defines this as never used - removing

File: src/Components/GameBoard.js [hunk 3]
```diff
@@ -1,17 +1,33 @@
...
 import Square from './Square';
 import * as utils from '../utils/functions';
+import CookieStorage from '../utils/CookieStorage';
```
Comment (debughug) [inline src/Components/GameBoard.js:4]: Importing the CookieStorage util

File: src/Components/GameBoard.js [hunk 4]
```diff
@@ -1,17 +1,33 @@
...
+
+const SAVE_DATA_NAME = 'game:tic-tac-toe';
+const DEFAULT_SAVE_STATE = Array(9).fill('');
```
Comment (debughug) [inline src/Components/GameBoard.js:7]: Declaring important consts. Such as the name of the save data cookie and default board value (empty)

File: src/Components/GameBoard.js [hunk 5]
```diff
@@ -1,17 +1,33 @@
...
   constructor(props) {
     super(props);
+    const squareArray = this.getSaveData();
+    const isXNext = utils.isXNext(squareArray);
+    this.state = { squareArray, isXNext };
+  }
+
+  setSaveData(squares) {
+    CookieStorage.setCookie(SAVE_DATA_NAME, JSON.stringify(squares));
+  }
```
Comment (debughug) [inline src/Components/GameBoard.js:19]: Triggered when a user clicks on a square. It takes the squares from the event saves it using the Cookie util

File: src/Components/GameBoard.js [hunk 6]
```diff
@@ -1,17 +1,33 @@
...
 
-    this.state = {
-      squareArray: Array(9).fill(null),
-      history: [],
-      xIsNext: true,
-    };
+  getSaveData() {
+    const SaveData = CookieStorage.getCookie(SAVE_DATA_NAME);
+    return SaveData === null ? DEFAULT_SAVE_STATE : JSON.parse(SaveData);
+  }
```
Comment (debughug) [inline src/Components/GameBoard.js:24]: Uses the Cookie util to get the stored data, if nothing is there then it'll use the const for an empty board

File: src/Components/GameBoard.js [hunk 7]
```diff
@@ -1,17 +1,33 @@
...
+
+  clearSaveData() {
+    CookieStorage.removeCookie(SAVE_DATA_NAME);
+    const squareArray = DEFAULT_SAVE_STATE;
+    const isXNext = utils.isXNext(squareArray);
+    this.setState({ squareArray, isXNext });
   }
```
Comment (debughug) [inline src/Components/GameBoard.js:31]: Clearing the board by removing the cookie data and setting the board to an empty state using the const. See how i've used it a bunch of times now? that's why it's a const.

File: src/Components/GameBoard.js [hunk 8]
```diff
@@ -22,12 +38,15 @@ export class GameBoard extends React.Component {
...
+    squares[num] = this.state.isXNext ? 'x' : 'o';
 
-    this.setState({
-      squareArray: squares,
-      xIsNext: !this.state.xIsNext,
-    });
+    this.setState(
+      {
+        squareArray: squares,
+        isXNext: !this.state.isXNext,
```
Comment (debughug) [inline src/Components/GameBoard.js:46]: Changed named to `isXNext` - I was told to always use the word IS before the rest of the name to indicate a boolean.

File: src/Components/GameBoard.js [hunk 9]
```diff
@@ -22,12 +38,15 @@ export class GameBoard extends React.Component {
...
+      },
+      this.setSaveData(squares),
```
Comment (debughug) [inline src/Components/GameBoard.js:48]: Having the `setSaveData` be used as the callback after the state has been updated

File: src/Components/GameBoard.js [hunk 10]
```diff
@@ -38,7 +57,8 @@ export class GameBoard extends React.Component {
     return (
       <div className="board-wrapper">
         <div className="hud">
-          <h3 className="turn-indicator">Current Player: {this.state.xIsNext ? 'X' : 'O'}</h3>
+          <h3 className="turn-indicator">Current Player: {this.state.isXNext ? 'X' : 'O'}</h3>
+          <button onClick={() => this.clearSaveData()}>Reset Game</button>
```
Comment (debughug) [inline src/Components/GameBoard.js:61]: Button to clear the board and cookie data

File: src/Styles/GameBoard.css [hunk 11]
```diff
@@ -9,7 +9,7 @@
   display: flex;
 }
 
-button {
+.square {
   font-size: 100px;
 }
```
Comment (debughug) [inline src/Styles/GameBoard.css:14]: Made it class based so it didn't mess with global buttons . e.g the reset board button

File: src/utils/functions.js [hunk 12]
```diff
@@ -1,20 +1,35 @@
...
+  squareArray.forEach((value) => {
+    if (value === 'x') {
+      xCount++;
+    } else if (value === 'o') {
+      oCount++;
+    }
+  });
+
+  return xCount < oCount;
+}
```
Comment (debughug) [inline src/utils/functions.js:35]: Check to see the count of X's and O's. Whoever has the least has the next turn.

**Comment by debughug** [comment]
@christulin  - Don't merge, we'll figure out how we merge things in as a group in next meet up

**Comment by joshuabradley012** [comment]
Looks clean. My gut says there is a cleaner way to do the `isXNext` logic. But I can't immediately think of how. And this works well.

`setState` with a callback is new to me, cool to see.

**Pull request closed without merging**